stdlib: add Array.init_matrix

gmevel opened the pull request:
This PR adds `Array.init_matrix` to the standard library, with the obvious semantics. It fills a gap because:

  1. a user expects to find this function, considering that there exist `Array.make` and `Array.init` and `Array.make_matrix`;
  2. just like `Array.make_matrix`, it reduces the risk of someone falling into the beginner’s trap of initializing an array with pointers to a shared mutable value.

I included a test.

Remaining questions:

  1. I would also add this function to `Float.Array`, but apparently there is no `Float.Array.make_matrix`. Is this intentional?
  2. The current docstring does not specify in which order the various invokations of `f` take place; should it? (I incline to think it should not, but this contrasts with the current docstring of `Array.init`.)
  3. This is my first PR. I am not sure what is the correct syntax for this docstring: `@since NEXT_OCAML_RELEASE`.

gmevel committed:
stdlib: add Array.init_matrix

xavierleroy reviewed (commented):
I think it's a good addition to the Array module.  I'm not 100% happy with the documentation, but have nothing better to suggest at this point.

gasche commented on testsuite/tests/lib-array/test_array.ml:
```diff
@@ -177,3 +177,16 @@ let a : int array =
+
+let a = Array.init_matrix 2 3 (fun i j -> ref (10*i+j));;
+a.(0).(0) := 99;;
+a (* other cells are unchanged *);;
```
This test of absence of unexpected sharing makes much more sense for `make_matrix`, which I suppose is not previously tested somewhere in this file? If so, I would suggest writing a test for `make_matrix` that does this modification check, and a usual test for `init_matrix` that just makes it evident that the result is different in all positions.

gasche reviewed (approved):
I agree that the function is reasonable and the implementation is fine.

gasche commented:
cc @nojb, who can tell us whether we should indeed add `Floatarray.{make,init}_matrix`.

yallop commented:
I'm in favour of adding this, but I'd prefer an implementation that went all in on either efficiency or elegance, e.g. 

```ocaml
let init_matrix sx sy f =
  init sx @@ fun x ->
  init sy @@ fun y ->
  f x y
```

 (or [perhaps one day](https://github.com/ocaml-flambda/ocaml-jst/pull/46) `let init_matrix3 sx sy f =  [| [|f x y for y = 0 to sy-1|] for x = 0 to sx-1 |]`)


Also, a documentation nit:

> ```
>  @raise Invalid_argument if [dimx] or [dimy] is negative or
> ```

If `dimx` is `0` and `dimy` is negative, it doesn't raise.

gasche commented:
I also found the implementation a bit odd, until I realized that it is adapted from `make_matrix` above which is structured in the same way.

gmevel committed:
Array: inline init in init_matrix

gmevel committed:
Array: add test for make_matrix, simplify test for init_matrix

gmevel commented:
Indeed I would have written the elegant version but I figured that maybe `make_matrix` was written that way for performance reasons, so I mimicked its code. I guess what @yallop is suggesting is that, if opting for ugly/efficient code, I should also inline the call to `init` as well and thus avoid the closure? I just pushed that change.

Regarding the doc nitpick (that part was copy-pasted from `make_matrix`): good catch, however I’m not sure whether we want to specify that `dimx = 0 && dimy < 0` is allowed. It means a more complex sentence in doc, I’m not sure it has practical benefits, and it could be regarded an accident of the implementation. Case in point, my reworking of the implementation has changed the behavior in this corner case (it now checks `dimy < 0` *before* `create`-ing the outer array; that’s not necessary, we could also swap these operations).

nojb commented on stdlib/array.ml:
```diff
@@ -63,6 +63,20 @@ let make_matrix sx sy init =
+let init_matrix sx sy f =
+  if sy < 0 then invalid_arg "Array.init_matrix";
+  let res = create sx [||] in
+  if sy > 0 then begin
```
This optimization (avoiding the loop if `sy = 0`) is not present in `make_matrix`. Not opposed to it, but I would rather have code with the same shape in both `init_matrix` and `make_matrix`.

gmevel replied:
It’s not just an optimization: without this test, we would evaluate `f x 0` for all `x` even if `sy = 0`. This is the meaning of the comment about #6575 in `init`. I made that comment more explicit, and added a new one in `init_matrix`.

I aligned the implementation of `make_matrix` with that of `init_matrix` as you suggested.

nojb commented:
> cc @nojb, who can tell us whether we should indeed add `Floatarray.{make,init}_matrix`.

I looked in LexiFi's codebase, there are a handful of occurrences of this function. So I would be in favour of adding them.

gmevel committed:
Array: make_matrix now raises if dimx = 0 && dimy < 0

This conforms to the documented behavior,
and aligns `make_matrix` with the new `init_matrix`.

gmevel committed:
Array: make_matrix now skips loop if dimx = 0

This aligns with the implementation of `init_matrix`, where this
short-circuit is not just an optimization, but a required test.

gmevel committed:
Float.Array: add {make,init}_matrix

gmevel commented:
I thus added `Float.Array.{make,init}_matrix`.

Regarding the edge case `dimx = 0 && dimy < 0` found by @yallop: while [addressing a reviewing comment](https://github.com/ocaml/ocaml/pull/12455#discussion_r1285071102) by @nojb, I aligned the behavior of `make_matrix` with that of `init_matrix`: now, both functions raise `Invalid_argument` (as do the new functions in `Float.Array`). I added a breaking change mention in the changelog. It has been the documented specification since 1999 (commit 10270af).

gasche commented:
I approved this already. If we want to accept this PR, we need a second approver -- because it is stdlib. In the past I have felt uneasy about being invited to second-approve because it is hard to say "no". So now I am hesitating to name names.

nojb commented on stdlib/array.mli:
```diff
@@ -91,6 +91,21 @@ val make_matrix : int -> int -> 'a -> 'a array array
+   If the return type of [f] is [float],
+   then the maximum size is only [Sys.max_array_length / 2].
+
+   @since NEXT_OCAML_RELEASE *)
```
Here (and below), `NEXT_OCAML_RELEASE` needs to be replaced by `5.2`.

gmevel replied:
Thanks! Is it to be done manually, or is there some tooling for this?

nojb replied:
Unless I missed some recent change, I think it has to be done by hand.

gmevel replied:
Done, rebased, and I also mentioned you as reviewer for the breaking change entry.

nojb reviewed (approved):
LGTM

gasche commented:
The CI has a failure of memory-model/forbidden.ml, which is unrelated.
